// @hint/hint-axe/dist/src/meta.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const category_1 = require("hint/dist/src/lib/enums/category");
const hint_scope_1 = require("hint/dist/src/lib/enums/hint-scope");
const meta = {
    docs: {
        category: category_1.Category.accessibility,
        description: 'Runs axe-core tests in the target',
        name: 'axe accessibility check'
    },
    id: 'axe',
    schema: [{
            additionalProperties: false,
            properties: {
                rules: {
                    patternProperties: {
                        '^.+$': {
                            additionalProperties: false,
                            properties: { enabled: { type: 'boolean' } },
                            required: ['enabled'],
                            type: 'object'
                        }
                    },
                    type: 'object'
                },
                runOnly: {
                    additionalProperties: false,
                    properties: {
                        type: { type: 'string' },
                        values: {
                            items: { type: 'string' },
                            minItems: 1,
                            type: 'array',
                            uniqueItems: true
                        }
                    },
                    type: 'object'
                }
            }
        }],
    scope: hint_scope_1.HintScope.any
};
exports.default = meta;
